shell/shared/grain.js

```diff
@@ -411,7 +411,11 @@ Router.map(function () {
         result.appOrigin = document.location.protocol + "//" + makeWildcardHost(session.hostId);
         setCurrentSessionId(session._id, result.appOrigin, grainId);
         result.sessionId = session._id;
-        result.path = encodeURIComponent("/" + (self.params.path || ""));
+        var currentPath = Iron.Location.get().path;
+        var rootPath = "/grain/" + grainId;
+        var grainPath = currentPath.slice(rootPath.length);
+        result.path = encodeURIComponent(grainPath);
+        result.hash = window.location.hash || "";
```

Comment:
I... don't know if this will lead to escaping problems in browsers. I think my brain capacity has been exceeded.